src/Parsers/tests/gtest_format_hiliting.cpp

```diff
@@ -0,0 +1,303 @@
+#include <unordered_set>
+
+#include <Parsers/IAST.h>
+#include <Parsers/ParserQuery.h>
+#include <Parsers/parseQuery.h>
+#include <gtest/gtest.h>
+#include <Common/StackTrace.h>
+
+std::string hilite(const std::string & s, const char * hilite_type)
+{
+    std::stringstream ss;
+    ss << hilite_type;
+    ss << s;
+    ss << DB::IAST::hilite_none;
+    return ss.str();
+}
+
+std::string keyword(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_keyword);
+}
+
+std::string identifier(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_identifier);
+}
+
+std::string alias(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_alias);
+}
+
+std::string op(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_operator);
+}
+
+std::string function(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_function);
+}
+
+std::string substitution(const std::string & s)
+{
+    return hilite(s, DB::IAST::hilite_substitution);
+}
+
+std::vector<const char *> HILITES =
+    {
+        DB::IAST::hilite_keyword,
+        DB::IAST::hilite_identifier,
+        DB::IAST::hilite_function,
+        DB::IAST::hilite_operator,
+        DB::IAST::hilite_alias,
+        DB::IAST::hilite_substitution,
+        DB::IAST::hilite_none
+    };
+
+[[maybe_unused]] const char * consume_hilites(const char ** it)
+{
+    const char * last_hilite = nullptr;
+    while (true)
+    {
+        bool changed_hilite = false;
+        for (const char * hilite : HILITES)
+        {
+            if (std::string_view(*it).starts_with(hilite))
+            {
+                *it += strlen(hilite);
```

Comment:
For your big refactoring: We could store the hilite strings in IAST as std::string_view, e.g.

```cpp
static constexpr std::string_view IAST::hilite_none = "\033[0m";
```

This will enable to call `.size()` on them w/o having to resort to a rather unaesthetic C function (l. 69).

Reply:
Ah, nice observation, I'll see if I'll be able to fit it in, thank you!

Reply:
True.

About performance: I'd say anything that does not run in a hot loop is not (obviously) performance critical. I doubt that the number of formatted queries per second is a reasonable metric. And `std::string_view` is really only a pointer + size and as fast (if not faster for things like `.size()`) than a plain `const char *`.

Reply:
Agree, formatted queries per second doesn't make sense.

Reply:
Why do you suggest having it as a std::string_view rather than just std::string?